### PR TITLE
Fix incorrect results with cloneWithType for finitetype integer sub types arrays

### LINKED ISSUE #30301: `cloneWithType` doesn't work as expected with finite types of numeric values
**Description:**
$title.

**Steps to reproduce:**
```ballerina
import ballerina/io;

type IntOneOrTwo 1|2;
type IntTwoOrThree 2|3;

type FloatOneOrTwo 1.0|2.0;

public function main() {
    int[] x = [1, 2];

    IntOneOrTwo[]|error a = x.cloneWithType(); // Should succeed, but fails atm.
    io:println(a is error); // true

    (IntOneOrTwo|IntTwoOrThree)[]|error b = x.cloneWithType(); // Should succeed, but fails atm.
    io:println(b is error); // true

    FloatOneOrTwo[]|error c = x.cloneWithType(); // Should succeed, but fails atm.
    io:println(a is error); // true

    int[] y = [1, 2, 3, 4];
    (IntOneOrTwo|IntTwoOrThree)[]|error d = y.cloneWithType(); // Should continue to fail.
    io:println(b is error); // true

    (IntOneOrTwo|FloatOneOrTwo)[]|error e = x.cloneWithType(); // Should continue to fail.
    io:println(b is error); // true
}
```

**Affected Versions:**
slalpha5

### LINKED ISSUE #31220: `cloneWithType()` fails for Integer sub types numeric conversion
**Description:**
This happens due to `isNumericType()` check is returning wrong value for int subtypes. 
https://github.com/ballerina-platform/ballerina-lang/blob/ecaec7d23f56082e80a02bcbd6327963dca21268/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java#L2214

We need to check the similar cases in other places as well. (`isValueType()`)

**Steps to reproduce:**
```ballerina
type IntSubTypeArray int:Signed32[];
float[] a1 = [1.23];
int:Signed32[] a2e = checkpanic a1.cloneWithType(IntSubTypeArray);
```
**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #30301
Fixes #31220

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by warunalakshitha — COMMENTED** [review]
LGTM

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 1]
```diff
@@ -2369,20 +2369,26 @@ private static boolean checkIsLikeArrayType(Object sourceValue, BArrayType targe
                         return true;
                     }
 
-                    if (targetTypeElementType.getTag() != TypeTags.UNION_TAG) {
+                    if (targetTypeElementType.getTag() != TypeTags.UNION_TAG &&
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:2372]: Shall we extract `targetTypeElementType.getTag()` out to a variable?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java [hunk 2]
```diff
@@ -236,21 +239,21 @@ static boolean isConvertibleToFloatingPointTypes(Object value) {
     }
 
     // TODO: return only the first matching type
-    public static List<Type> getConvertibleTypes(Object inputValue, Type targetType) {
+    public static Set<Type> getConvertibleTypes(Object inputValue, Type targetType) {
         return getConvertibleTypes(inputValue, targetType, new ArrayList<>());
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:243]: The following still doesn't work.

```ballerina
    float[] x = [1, 2];
    (int|byte)[] y = checkpanic x.cloneWithType();
```

Comment (MaryamZi) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:243]: Need to check with combinations like int and int subtypes too.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java [hunk 3]
```diff
@@ -266,6 +269,18 @@ static boolean isConvertibleToFloatingPointTypes(Object value) {
                     convertibleTypes.add(matchingType);
                 }
                 break;
+            case TypeTags.FINITE_TYPE_TAG:
+                for (Object valueSpaceItem : ((BFiniteType) targetType).valueSpace) {
+                    Type inputValueType = TypeChecker.getType(inputValue);
+                    Type valueSpaceItemType = TypeChecker.getType(valueSpaceItem);
+                    if (inputValue == valueSpaceItem && inputValueType.getTag() == valueSpaceItemType.getTag()) {
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:276]: Can you please share an example for when the value is the same but the type tag is different?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java (hunk 2 above)
Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:244]: Addressed with https://github.com/ballerina-platform/ballerina-lang/pull/30710/commits/48f5e12c00c98975124f5e2a952137570e4d8579

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java (hunk 1 above)
Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java]: This piece of code has been removed with new commits

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java (hunk 3 above)
Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java]: There should be no possible cases, just added for a safe check but it looks like redundant. Removed with a later commit.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 4]
```diff
@@ -2390,7 +2390,8 @@ private static boolean checkIsLikeArrayType(Object sourceValue, BArrayType targe
                     return targetNumericTypes.size() == 1;
                 }
 
-                if (isNumericType(targetTypeElementType) && targetTypeElementType.getTag() != TypeTags.BYTE_TAG) {
+                if (isNumericType(targetTypeElementType) && targetTypeElementType.getTag() != TypeTags.BYTE_TAG &&
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:2393]: If we just check targetTypeElementType == FLOAT_TAG || targetTypeElementType == DECIMAL_TAG

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java [hunk 5]
```diff
@@ -428,8 +443,29 @@ private static boolean isConvertibleToArrayType(Object sourceValue, BArrayType t
...
+            if (convertibleTypes.size() != 1 && !convertibleTypes.contains(TypeChecker.getType(source.get(i)))
+                    && !hasIntegerSubTypes(convertibleTypes)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static boolean hasIntegerSubTypes(Set<Type> convertibleTypes) {
+        if (convertibleTypes.size() < 2) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:464]: Do we need this check?

**Comment by gabilang** [comment]
There are some other known issues related to `cloneWithType` with tuples. Created an [issue](https://github.com/ballerina-platform/ballerina-lang/issues/31747) for it.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java (hunk 5 above)
Comment (warunalakshitha) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java:464]: convertibleTypes.size() != 1  and empty is already checked in previous conditions.

File: langlib/lang.value/src/main/java/org/ballerinalang/langlib/value/CloneWithType.java [hunk 6]
```diff
@@ -107,16 +108,23 @@ private static Object convert(Object value, Type targetType, List<TypeValuePair>
...
         }
-        List<Type> convertibleTypes;
+        Set<Type> convertibleTypes;
         convertibleTypes = TypeConverter.getConvertibleTypes(value, targetType);
+
+        Type sourceType = TypeChecker.getType(value);
         if (convertibleTypes.isEmpty()) {
             throw createConversionError(value, targetType);
-        } else if (!allowAmbiguity && convertibleTypes.size() > 1) {
+        } else if (!allowAmbiguity && convertibleTypes.size() > 1 && !convertibleTypes.contains(sourceType) &&
```
Comment (warunalakshitha) [inline langlib/lang.value/src/main/java/org/ballerinalang/langlib/value/CloneWithType.java:117]:
```suggestion
        } else if (!allowAmbiguity && convertibleTypes.size() > 2 && !convertibleTypes.contains(sourceType) &&
```

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java (hunk 5 above)
Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java]: Yeah, `convertibleTypes.size() < 2` is kind of repeated check, that is not needed here. But we need to preserve the `convertibleTypes.size()  == 0` check for other places. Moved that check to outside of this method. 
Addressed with https://github.com/ballerina-platform/ballerina-lang/pull/30710/commits/688ab907b7244cd39e24c251849c329e7f7fbe24

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java (hunk 4 above)
Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java]: Addressed with https://github.com/ballerina-platform/ballerina-lang/pull/30710/commits/688ab907b7244cd39e24c251849c329e7f7fbe24

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**